draft-ietf-quic-transport.md

```diff
@@ -543,6 +544,17 @@ the server supports, as shown below.
 ~~~
 {: #version-negotiation-format title="Version Negotiation Packet"}
 
+The value in the Unused field is selected randomly by the server. The Connection
```

Comment:
This will make it impossible to determine the packet type by looking at the type byte.

Reply:
That is true, but you can look at the type byte and version fields together to determine it. Also note that this is only sent on the server -> client direction, so you're not going to hit boxes that need to scale this sort of check.

Reply:
True, but why grease it at all? Note that with this proposal, you can't trust the long header type byte **at all**, since every reserved value could also just be the random value of the version negotiation packet.
This will make parsing long header packets quite painful, you will always have to look at the other fields to determine the true packet byte.
For middle boxes, the version negotiation packet will be clearly distinguishable anyway, so I don't see the need to play tricks with the type byte.

Reply:
Removing version negotiation from consideration makes it possible to grease the type field completely.  Creating a carve-out for version negotiation makes it a little harder.

Long header packets are rare, so it seems like hitting the slow path is tolerable.

Reply:
We could exclude the code points defined for Long Header types from the random values that can be used for the greased type byte of the version negotiation packet. That way, we get greasing and easy parsing at the same time.

Reply:
"the other fields" => version.  Since the type byte is version-specific, you should always confirm that the version is what you expect before attempting to interpret the type byte.  (Should we perhaps mention this in the details of handling incoming packets?)

Reply:
Ok, I didn't think about multiple version. Since the type byte version dependent, I'd argue that we shouldn't grease it at all for version negotiation packets.This would simplify the parsing logic, since you could identify the version negotiation packet by just looking at it's type byte (which is a QUIC invariant). For all other packets, you'd know the version anyway (because they're received after version negotiation).
If we want to grease, we could define a greasing mechanism for long header packets.

Reply:
But the greasing mechanism for long-header packets would likely be version-specific.  I think the point here is that "version 0" doesn't actually define any semantics for the Type field because it doesn't define any types; it defines a single type, and the value in that field is irrelevant once you've seen the version.

So the parsing logic would be:

```
if( in_handshake ) {
  switch(packet.version) {
    what_i_sent:  switch(packet.type) { ... }
    0:  version_negotiation(packet)
    default:  return;
  }
else {
  if(packet.version != connection.version) return;
  ...
}
```

Reply:
The principle I'm advocating here (as well as in #426) is that a packet header should be identifiable by just looking at its type byte. So the parsing logic would be:

```
ParseHeader(our_version) {
  if(packet.type_byte == kTypeByteVersionNegotiation) {
    // parse version negotiation packet
  }
  else {
    // version dependent parsing of the packet header (version is our_version)
  }
}
```

Note that the **only** thing in the packet you have to look at to correctly parse the header is the type byte.

Reply:
We should move this to #963, since this is a discussion about whether the issue in #963 is a problem or a feature.

Reply:
Thanks. I must have missed that issue. I'll write up my argument there.

Reply:
This looks like a gratuitous change. What exactly is the benefit of not having a version renegotiation packet?

Reply:
@huitema: can we discuss on #963 ?